src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpHandlerResourceV3Provider.cs

```diff
@@ -105,17 +125,41 @@ public CredentialPromptWebRequestHandler(IWebProxy proxy, ICredentials credentia
                         if (ProxyAuthenticationRequired(ex) &&
                             HttpHandlerResourceV3.PromptForProxyCredentials != null)
                         {
-                            // prompt use for proxy credentials.
-                            var credentials = await HttpHandlerResourceV3
-                                .PromptForProxyCredentials(request.RequestUri, Proxy, cancellationToken);
+                            ICredentials currentCredentials = Proxy.Credentials;
 
-                            if (credentials == null)
+                            try
                             {
-                                throw;
+                                await _credentialPromptLock.WaitAsync();
+
+                                // Check if the credentials have already changed
+                                if (!object.ReferenceEquals(currentCredentials, Proxy.Credentials))
+                                {
+                                    continue;
+                                }
+
+                                // Limit the number of retries
+                                _authRetries++;
```

Comment:
yes, there have been several bugs where it was reported that this went into an infinite loop.